app/src/ui/lib/avatar.tsx

```diff
@@ -15,74 +15,134 @@ interface IAvatarProps {
+  if (user === undefined) {
+    return
   }
 
-  public shouldComponentUpdate(
-    nextProps: IAvatarProps,
-    nextState: IAvatarState
-  ) {
-    if (nextState.dataUrl !== this.state.dataUrl) {
-      return true
-    }
-
-    if (nextProps.user && this.props.user) {
-      if (
-        nextProps.user.avatarURL !== this.props.user.avatarURL ||
-        nextProps.user.email !== this.props.user.email ||
-        nextProps.user.name !== this.props.user.name ||
-        nextProps.title !== this.props.title
-      ) {
-        return true
+  const { email, endpoint, avatarURL } = user
+
+  if (endpoint === getDotComAPIEndpoint()) {
+    // The avatar urls returned by the API doesn't come
+    // with a size parameter, they default to the biggest
+    // size we need on GitHub.com which is usually much bigger
+    // than what desktop needs so we'll set a size explicitly.
+    if (avatarURL !== undefined) {
+      try {
+        const url = new URL(avatarURL)
+        url.searchParams.set('s', `${size}`)
+
+        yield url.toString()
+      } catch (e) {
+        // This should never happen since URL#constructor
+        // only throws for invalid URLs which we can expect
+        // the API to not give us
+        yield avatarURL
       }
     }
+  } else if (endpoint !== null) {
+    // We're dealing with a repository hosted on GitHub Enterprise Server
+    // so we're unable to get to the avatar by requesting the avatarURL due
+    // to the private mode (see https://github.com/desktop/desktop/issues/821).
+    // So we have no choice but to fall back to gravatar for now.
+    yield generateGravatarUrl(email, size)
+    return
+  }
 
-    if (!nextProps.user && !this.props.user) {
-      return nextProps.title !== this.props.title
+  // Are we dealing with a GitHub.com stealth/anonymous email address in
+  // either legacy format:
+  //  niik@users.noreply.github.com
+  //
+  // or the current format
+  //  634063+niik@users.noreply.github.com
+  //
+  // If so we unfortunately can't rely on the GitHub avatar endpoint to
+  // deliver a match based solely on that email address but luckily for us
+  // the avatar service supports looking up a user based either on user id
+  // of login, user id being the better option as it's not affected by
+  // account renames.
+  const stealthEmailMatch = StealthEmailRegexp.exec(email)
+
+  if (stealthEmailMatch) {
+    const [, userId, login] = stealthEmailMatch
+    if (userId !== undefined) {
+      yield `${avatarEndpoint}/u/${encodeURIComponent(userId)}?s=${size}`
+    } else {
+      yield `${avatarEndpoint}/${encodeURIComponent(login)}?s=${size}`
     }
-
-    return false
   }
 
-  public async componentWillMount() {
-    const dataUrl = await fetchAvatarUrl(DefaultAvatarURL, this.props.user)
+  // The /u/e endpoint above falls back to gravatar (proxied)
+  // so we don't have to add gravatar to the fallback.
+  yield `${avatarEndpoint}/u/e?email=${encodeURIComponent(email)}&s=${size}`
```

Comment:
I think we should still add the gravatar fallback for non-github or local repositories (in the case that our endpoint does not proxy to gravatar for non-existant users, which may be quite probable), what do you think?

(I think ideally we should not do the call to our endpoint on non-github repositories but keep it for local repositories, but that may be hard to do without access to the whole `Repository` object).

Reply:
Given that GitHub's avatar endpoint will always fall back to gravatar (and if that fails it'll return a generic identicon) I think adding one more fallback is only gonna add to the loading time before we show the no-avatar-avatar.

> (I think ideally we should not do the call to our endpoint on non-github repositories but keep it for local repositories, but that may be hard to do without access to the whole `Repository` object).

I agree with that reasoning but just as you say, it's hard to do without access to the list of remotes.

Reply:
Cool, let's leave it like this then!